File: android/app/src/main/java/dev/ethone/app/service/FocusManager.kt
package dev.ethone.app.service

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.compose.ui.graphics.Color
import dev.ethone.app.ui.theme.EthoneAmber
import dev.ethone.app.ui.theme.EthoneCyan
import dev.ethone.app.ui.theme.EthoneEmerald
import dev.ethone.app.ui.theme.EthoneViolet
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch

enum class FocusPreset(val label: String, val durationMinutes: Int, val tint: Color) {
    Pomodoro("Pomodoro", 25, EthoneEmerald),
    DeepWork("Deep Work", 50, EthoneViolet),
    Sprint("Sprint", 10, EthoneAmber),
    ShortBreak("Pause Courte", 5, EthoneCyan)
}

object FocusManager {
    private val scope = CoroutineScope(Dispatchers.Main)
    private var timerJob: Job? = null

    var currentPreset by mutableStateOf(FocusPreset.Pomodoro)
        private set

    var isRunning by mutableStateOf(false)
        private set

    var isPaused by mutableStateOf(false)
        private set

    var totalSeconds by mutableIntStateOf(25 * 60)
        private set

    var remainingSeconds by mutableIntStateOf(25 * 60)
        private set

    var completedSessions by mutableIntStateOf(0)
        private set

    val progress: Float
        get() = if (totalSeconds > 0) (totalSeconds - remainingSeconds).toFloat() / totalSeconds.toFloat() else 0f

    val formattedTime: String
        get() {
            val minutes = remainingSeconds / 60
            val seconds = remainingSeconds % 60
            return String.format("%02d:%02d", minutes, seconds)
        }

    fun selectPreset(preset: FocusPreset) {
        if (isRunning) return
        currentPreset = preset
        totalSeconds = preset.durationMinutes * 60
        remainingSeconds = totalSeconds
    }

    fun start() {
        if (isRunning) return
        isRunning = true
        isPaused = false

        timerJob = scope.launch {
            while (isActive && remainingSeconds > 0) {
                delay(1000)
                if (!isPaused) {
                    remainingSeconds--
                }
            }
            if (remainingSeconds <= 0) {
                completeSession()
            }
        }
    }

    fun pause() {
        if (!isRunning || isPaused) return
        isPaused = true
    }

    fun resume() {
        if (!isRunning || !isPaused) return
        isPaused = false
    }

    fun stop() {
        timerJob?.cancel()
        timerJob = null
        isRunning = false
        isPaused = false
        remainingSeconds = totalSeconds
    }

    private fun completeSession() {
        stop()
        completedSessions++
    }
}

File: ios/App/App/Services/FocusManager.swift
import Foundation
import Combine
import SwiftUI

public enum FocusMode: String, CaseIterable, Identifiable {
    case pomodoro = "Pomodoro"
    case deepWork = "Deep Work"
    case sprint = "Sprint"
    case shortBreak = "Pause Courte"
    case longBreak = "Pause Longue"

    public var id: String { rawValue }

    public var defaultDurationSeconds: Int {
        switch self {
        case .pomodoro: return 25 * 60
        case .deepWork: return 50 * 60
        case .sprint: return 10 * 60
        case .shortBreak: return 5 * 60
        case .longBreak: return 15 * 60
        }
    }

    public var icon: String {
        switch self {
        case .pomodoro: return "timer"
        case .deepWork: return "brain"
        case .sprint: return "bolt.fill"
        case .shortBreak: return "cup.and.saucer.fill"
        case .longBreak: return "bed.double.fill"
        }
    }

    public var tintColor: Color {
        switch self {
        case .pomodoro: return ETHTheme.emerald
        case .deepWork: return ETHTheme.violet
        case .sprint: return ETHTheme.amber
        case .shortBreak: return ETHTheme.cyan
        case .longBreak: return ETHTheme.teal
        }
    }
}

@MainActor
public final class FocusManager: ObservableObject {
    public static let shared = FocusManager()

    @Published public var currentMode: FocusMode = .pomodoro
    @Published public var isActive: Bool = false
    @Published public var isPaused: Bool = false
    @Published public var totalSeconds: Int = 25 * 60
    @Published public var remainingSeconds: Int = 25 * 60
    @Published public var completedSessions: Int = 0

    private var timer: Timer?

    public init() {
        self.totalSeconds = currentMode.defaultDurationSeconds
        self.remainingSeconds = totalSeconds
    }

    public var progress: Double {
        guard totalSeconds > 0 else { return 0 }
        return Double(totalSeconds - remainingSeconds) / Double(totalSeconds)
    }

    public var formattedTime: String {
        let minutes = remainingSeconds / 60
        let seconds = remainingSeconds % 60
        return String(format: "%02d:%02d", minutes, seconds)
    }

    public func selectMode(_ mode: FocusMode) {
        guard !isActive else { return }
        currentMode = mode
        totalSeconds = mode.defaultDurationSeconds
        remainingSeconds = totalSeconds
        HapticManager.shared.selection()
    }

    public func start() {
        guard !isActive else { return }
        isActive = true
        isPaused = false
        HapticManager.shared.medium()

        if #available(iOS 26.0, *) {
            FocusLiveActivityManager.shared.start(
                preset: currentMode.rawValue,
                totalDuration: TimeInterval(totalSeconds),
                endDate: Date().addingTimeInterval(TimeInterval(remainingSeconds))
            )
        }

        timer = Timer.scheduledTimer(withTimeInterval: 1.0, repeats: true) { [weak self] _ in
            guard let self = self else { return }
            Task { @MainActor in
                self.tick()
            }
        }
    }

    public func pause() {
        guard isActive, !isPaused else { return }
        isPaused = true
        timer?.invalidate()
        timer = nil
        HapticManager.shared.light()

        if #available(iOS 26.0, *) {
            FocusLiveActivityManager.shared.update(
                progress: progress,
                remaining: TimeInterval(remainingSeconds),
                isRunning: false
            )
        }
    }

    public func resume() {
        guard isActive, isPaused else { return }
        isPaused = false
        HapticManager.shared.light()

        if #available(iOS 26.0, *) {
            FocusLiveActivityManager.shared.update(
                progress: progress,
                remaining: TimeInterval(remainingSeconds),
                isRunning: true
            )
        }

        timer = Timer.scheduledTimer(withTimeInterval: 1.0, repeats: true) { [weak self] _ in
            guard let self = self else { return }
            Task { @MainActor in
                self.tick()
            }
        }
    }

    public func stop() {
        timer?.invalidate()
        timer = nil
        isActive = false
        isPaused = false
        remainingSeconds = totalSeconds
        HapticManager.shared.heavy()

        if #available(iOS 26.0, *) {
            FocusLiveActivityManager.shared.end()
        }
    }

    private func tick() {
        guard remainingSeconds > 0 else {
            completeSession()
            return
        }
        remainingSeconds -= 1

        if #available(iOS 26.0, *) {
            FocusLiveActivityManager.shared.update(
                progress: progress,
                remaining: TimeInterval(remainingSeconds),
                isRunning: !isPaused
            )
        }
    }

    private func completeSession() {
        timer?.invalidate()
        timer = nil
        isActive = false
        isPaused = false
        completedSessions += 1
        remainingSeconds = totalSeconds
        HapticManager.shared.success()

        if #available(iOS 26.0, *) {
            FocusLiveActivityManager.shared.end()
        }
    }
}
